php/libraries/Candidate.class.inc

```diff
@@ -680,115 +679,272 @@ class Candidate
     */
     static function _generateCandID(): int
     {
-        // get a database connection
-        $factory = NDB_Factory::singleton();
-        $db      = $factory->database();
-
-        // generate a candidate id
-        $candID = random_int(CANDIDATE_MIN_CANDID, CANDIDATE_MAX_CANDID);
-
-        // make sure the cand id is not used
-        while ($db->pselectOne(
-            "SELECT count(CandID) FROM candidate WHERE CandID=:CaID",
-            array('CaID' => $candID)
-        ) > 0) {
-            // pick a new candidate id
-            $candID = random_int(CANDIDATE_MIN_CANDID, CANDIDATE_MAX_CANDID);
-        }
-
-        return $candID;
+        $existingIDs = \Candidate::_getExistingIDs('CandID');
```

Comment:
This introduces a race condition which can result in multiple simultaneous requests generating the same CandID.

Reply:
I added an exception in the calling code to go off when a newly-generated PSCID or CandID is not unique.

Reply:
That doesn't address the race condition. 2 different requests independently get a list of IDs here, and then once one inserts something the list that the other one is verifying against is stale.

Reply:
Do you have a suggestion of how to approach this? I'm not sure how to deal with this either than querying the DB like in the exception case I added.

Reply:
@johnsaigle As CandID is the primary key, trying to insert a CandID that is already used will generate an error. You can do a try catch (or while loop) and if the error is "existing primary key", increment CandID and/or PSCID.

Not perfect as you might have to increment a few time (would be rare but if 10 persons are adding new candidates, it might append. Also, when increasing the CandID, you risk going over the max value and/or exhausting the bank of possible values.

Unless you want to use a lock table (not the candidate table!) or other lock mechanism.
@driusan  which method would you prefer?